### PR TITLE
Refactor/sweep runner

### PR DESCRIPTION
**Pull request by jboreiko**
No release notes required.  This is primarily a testing addition and will be followed up with additional parts of a refactor to simplify the runner logic.  This can then be followed with an addition of column paging for sweep.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/710)

### CODE REVIEW

**Comment by gsheasby** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 129 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX1DuofekPKIJKBvQN:-KNX1Dup0SNrEfPjdIB4:2105183550) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L129)):*
> ```Java
>         // (2) force old writers to retry (note that we must roll back any uncommitted transactions that
>         //     we encounter
>         SweepStrategy sweepStrategy = MoreObjects.firstNonNull(sweepStrategyManager.get().get(tableRef), SweepStrategy.CONSERVATIVE);
> ```

Does Collections not cover this?

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 277 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX3-YLviOnyfxZ4Nz_:-KNX3-YLviOnyfxZ4Nza:-1213335446) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L277)):*
> ```Java
>     }
> ```

Double blank line

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 318 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX33yb6rRhanxHOSt1:-KNX33yb6rRhanxHOSt2:-1213019852) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L318)):*
> ```Java
>         keyValueService.delete(tableRef, cellTsPairsToSweep);
>     }
> ```

more random whitespace

---

*[atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java, line 2 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX3BjAw1CRJvdFaVAK:-KNX3BjAw1CRJvdFaVAL:-1077879633) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java#L2)):*
> ```Java
> /**
>  * Copyright 2015 Palantir Technologies
> ```

2016 - you can change the copyright template to insert the current year.

---

*[atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java, line 55 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX3cHm0dUBaCYYGYHr:-KNX3cHm0dUBaCYYGYHs:918701004) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java#L55)):*
> ```Java
>     private final KeyValueService mockKVS = mock(KeyValueService.class);
>     private final Follower mockFollower = mock(Follower.class);
>     private final SweepTaskRunnerImpl sweepTaskRunner = new SweepTaskRunnerImpl(null, mockKVS, null, null, null, null, ImmutableList.of(mockFollower));
> ```

BTW: this looks pretty nasty, but OK as long as we aren't throwing lots of nulls around in multiple places.

---

*[atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java, line 122 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX41jUoCTPtWsb3-v4:-KNX41jUoCTPtWsb3-v5:-2068977985) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java#L122)):*
> ```Java
>     @Test
>     public void noTimetampsAreFilteredOutWhenGettingTimestampsFromRowResultsInThoroughSweep() {
>         List<RowResult<Set<Long>>> cellsToSweep = ImmutableList.of(RowResult.of(SINGLE_CELL, ImmutableSet.of(Value.INVALID_VALUE_TIMESTAMP, 1L)));
> ```

What does `1L` mean here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/710)*

**Comment by SerialVelocity** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 129 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX1DuofekPKIJKBvQN:-KNX5ppXcsqvL4PrSI_Y:-883797925) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L129)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Does Collections not cover this?
</details>
Talked about this in chat. We want to leave it semantically equivalent until we look over whether values in the map can be null here.

---

*[atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java, line 122 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX41jUoCTPtWsb3-v4:-KNX5YL-5_lG4KpnMcAA:595143932) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java#L122)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> What does `1L` mean here?
</details>
1L because if I write just 1, then We get a `Set<Long, Integer>` which would create the wrong type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/710)*

**Comment by gsheasby** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 129 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX1DuofekPKIJKBvQN:-KNXAkuYMY87-lsYYVrd:1853604893) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L129)):*
<details><summary><i>Previously, SerialVelocity (Ben Grabham) wrote…</i></summary>
> Talked about this in chat. We want to leave it semantically equivalent until we look over whether values in the map can be null here.
</details>
I'd still prefer `sweepStrategyManager.get().getOrDefault(tableRef, SweepStrategy.CONSERVATIVE)` - it does exactly what we need, without depending on some library.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 128 \[r2\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNXAnsfzHWEQTQI4HLM:-KNXAnsfzHWEQTQI4HLN:-520406618) ([raw file](https://github.com/palantir/atlasdb/blob/36b6c837d1e1941be9076e16de9da26f3119e479/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L128)):*
> sweepStrategyManager.get().get(tableRef), SweepStrategy.CONSERVATIVE);

---

*[atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java, line 101 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNXBL3JLbFjpZNo5aYZ:-KNXBL3JLbFjpZNo5aY_:387827569) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java#L101)):*
> ```Java
> 
>     @Test
>     public void noValidTimestampsAreFilteredOutWhenGettingTimestampsFromRowResultsInThoroughSweep() {
> ```

What happened to this test? It looks different from the one it appears to have been replaced with.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/710)*

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java [hunk 1]
```diff
@@ -123,24 +126,19 @@ public SweepResults run(TableReference tableRef, int batchSize, @Nullable byte[]
...
+        SweepStrategy sweepStrategy = MoreObjects.firstNonNull(sweepStrategyManager.get().get(tableRef), SweepStrategy.CONSERVATIVE);
+        if (sweepStrategy == SweepStrategy.NOTHING) {
             // This sweep strategy makes transaction table truncation impossible
             return SweepResults.createEmptySweepResult(0L);
         }
-        if (startRow == null) {
-            startRow = new byte[0];
-        }
+
+        startRow = MoreObjects.firstNonNull(startRow, new byte[0]);
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java]: nit: just reference an `EMPTY_BYTE_ARRAY` as fallback

**Comment by SerialVelocity** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 129 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX1DuofekPKIJKBvQN:-KNXagl5sMuea0g1IYnv:-1203018840) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L129)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> I'd still prefer `sweepStrategyManager.get().getOrDefault(tableRef, SweepStrategy.CONSERVATIVE)` - it does exactly what we need, without depending on some library.
</details>
It's actually different semantically since `getOrDefault` will return `null` if `null` is in the `Map`. At the moment, we are keeping everything semantically the same and only adding tests. That is something we can change it to once we investigate into whether a `null` can ever exist within the map.

---

*[atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java, line 101 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNXBL3JLbFjpZNo5aYZ:-KNXa_xiqxWwdnz-DDch:-1492970274) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java#L101)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> What happened to this test? It looks different from the one it appears to have been replaced with.
</details>
? This is a new test. It is the same as the one above but using CONSERVATIVE instead of THOROUGH

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/710)*

**Comment by SerialVelocity** [comment]
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java, line 101 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNXBL3JLbFjpZNo5aYZ:-KNXipUV8W356oywEAW_:-1017657959) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/test/java/com.palantir.atlasdb.sweep/SweepTaskRunnerImplTest.java#L101)):*
<details><summary><i>Previously, SerialVelocity (Ben Grabham) wrote…</i></summary>
> ? This is a new test. It is the same as the one above but using CONSERVATIVE instead of THOROUGH
</details>
Oh, reviewable put this comment on the line additions. validTimestampsAreKeptWhenGettingTimestampsFromRowResultsInThoroughSweep and invalidTimetampsAreKeptWhenGettingTimestampsFromRowResultsInThoroughSweep cover this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/710)*

**Comment by SerialVelocity** [comment]
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 129 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX1DuofekPKIJKBvQN:-KNXk-W9sk0LeScEnwlV:-2111063677) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L129)):*
<details><summary><i>Previously, SerialVelocity (Ben Grabham) wrote…</i></summary>
> It's actually different semantically since `getOrDefault` will return `null` if `null` is in the `Map`. At the moment, we are keeping everything semantically the same and only adding tests. That is something we can change it to once we investigate into whether a `null` can ever exist within the map.
</details>
For reference, here's the default implementation:
```
    default V getOrDefault(Object key, V defaultValue) {
        V v;
        return (((v = get(key)) != null) || containsKey(key))
            ? v
            : defaultValue;
    }
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/710)*

**Comment by schlosna** [comment]
Reviewed 1 of 3 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/710)*

**Comment by gsheasby** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 129 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/710#-KNX1DuofekPKIJKBvQN:-KNXsQo5DCXZnXR3a3Jc:-162904695) ([raw file](https://github.com/palantir/atlasdb/blob/0e0f3b2a9264a10861d7938ab9d9eb4db415b85d/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L129)):*
<details><summary><i>Previously, SerialVelocity (Ben Grabham) wrote…</i></summary>
> For reference, here's the default implementation:
> ```
>     default V getOrDefault(Object key, V defaultValue) {
>         V v;
>         return (((v = get(key)) != null) || containsKey(key))
>             ? v
>             : defaultValue;
>     }
> ```
</details>
ok, LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/710#-:-KNXsqNHS4iM-dteZLn4:1417088721)*

**Pull request merged**